src/yara.ml

```diff
@@ -82,16 +82,69 @@ let finalize () =
 let finalize_exn () = Yr.unwrap @@ finalize ()
 
 module Rule = struct
-  type t = Yara_c.Yr_rule.t ptr
-
-  let get_identifier rule =
-    let rule = !@rule in
-    getf rule Yara_c.Yr_rule.identifier
+  type metadata =
+    | Int of int64
+    | Bool of bool
+    | String of string
+
+  type t = {
+    identifier : string;
+    namespace : string;
+    metadata : (string * metadata) list;
+  }
+
+  let collect_metadata metas =
+    let q = Queue.create () in
+    let ident meta = getf meta Yara_c.Yr_meta.identifier in
+    let rec loop metas =
+      if not (is_null metas) then (
+        let meta = !@metas in
+        let flags = getf meta Yara_c.Yr_meta.flags in
+        let last_in_rule = flags = Yara_c.Constants.meta_flags_last_in_rule in
+        let ident = ident meta in
+        match getf meta Yara_c.Yr_meta.type_ with
+        | Yara_c.Meta_type.Integer ->
+          let v = getf meta Yara_c.Yr_meta.integer in
+          Queue.add (ident, Int v) q;
+          if not last_in_rule then loop (metas +@ 1)
+        | String ->
+          let v = getf meta Yara_c.Yr_meta.string in
+          Queue.add (ident, String v) q;
+          if not last_in_rule then loop (metas +@ 1)
+        | Boolean ->
+          let v = getf meta Yara_c.Yr_meta.integer in
+          Queue.add
+            ( ident,
+              Bool
+                ( if v > 0L then
+                  true
+                else
+                  false
+                )
+            )
+            q;
+          if not last_in_rule then loop (metas +@ 1)
+        | Unknown _ -> loop (metas +@ 1)
```

Comment:
Not sure if we want to error here (exception), or silently keep iterating like this does now. I'm guessing we will never reach this branch as i'd be surprised if libyara doesn't error early if a rule attempts to use an invalid metadata type?

Reply:
Agreed. I think it might be better to perform this check, and raise an exception at the `Yara_c` layer where we initially see this value. That'll allow dropping the `Unknown` variant.

Reply:
I've removed the `Unknown` branch!